addon/mixins/callback-route-mixin.js

```diff
@@ -0,0 +1,69 @@
+import Ember from 'ember';
+
+const { inject: { service }, Mixin } = Ember;
+
+export default Mixin.create({
+  /**
+   The session service.
+
+   @property session
+   @readOnly
+   @type SessionService
+   @public
+   */
+  session: service('session'),
+
+  /**
+   The authenticator that should be used to authenticate the callback.
+
+   @property authenticator
+   @type String
+   @default null
+   @public
+   */
+  authenticator: null,
+
+  /**
+   The route that should be transitioned to on error.
+
+   @property routeOnError
+   @type String
+   @default null
+   @public
+   */
+  routeOnError: null,
+
+  /**
+   The route that should be transitioned to on success.
+
+   @property routeOnSuccess
+   @type String
+   @default null
+   @public
+   */
+  routeOnSuccess: null,
+
+  /**
+   Authenticates the passed authentication server response and transitions
+   to the `routeOnError` route with the error specified as query parameters
+   or to the `routeOnSuccess`route.
+
+   If the current windows is a pop-up it will be automatically closed on success.
+
+   @method activate
+   @public
+   */
+  activate() {
+    let authenticator = this.get('authenticator');
+
+    this.get('session').authenticate(authenticator).then(() => {
+      if (window.opener) {
+        return window.close();
+      }
```

Comment:
This won't result in any notification to the opener, meaning the opener would never know what the outcome of the authentication operation was. Implementing this in a proper way would require us to rebuild large parts of torii which I don't think is sth. we want to do.

I'd leave this out and leave it to people using this mixin to override the `activate` method if they want to have that behavior.